### ISSUE TITLE
[BUG] - Slowness

### ISSUE THREAD

**Issue by freijon** [zaneschepke/wgtunnel#324]
Since a while I notice horrible performance with Wireguard. A speedtest shows I can only reach 0.35 Mbps. I downgraded to 3.4.8 and 3.4.7 which solves the problem.

- Device: OnePlus Nord with /e/OS 2.2
- Android Version: 12
- App Version 3.5.0
- Doesn't happen with version: 3.4.8 or 3.4.7

**Comment by zaneschepke** [comment]
Hello! Could you share the model of your phone? I have been hearing reports of this on Redmi devices.

**Comment by freijon** [comment]
Sure. Added in description.

**Comment by crimsonthunder** [comment]
Same for me on Google Pixel 6.
Android 15 Public beta latest 
App Version 3.5.1

**Comment by mr-scrawley** [comment]
Also affected on:
- Samsung Galaxy S21
- OneUI 6.1 (Android 14)
- WG Tunnel 3.5.1

Rollback to 3.4.8 fixes restores performance to what the official wireguard app provides.

**Comment by hudcap** [comment]
New user here (great app btw - thanks!). I was only seeing an issue on mobile data, not wifi. I assumed it was a "me" issue, so I created a separate VPN for mobile data and tried changing the MTU, to no avail.
Came back here and found this - Amnezia solved it, so I assume I'm bitten by the same bug.

TL;DR this issue only affected me on mobile data (Verizon via Red Pocket)

**Comment by GeneralTDog** [comment]
For Protocol:
Same issue here (OnePlus 8T on lineage_kebab-userdebug 14 AP2A.240705.005 3d40349489 with wg-tunnel 3.5.1)

Using Amnezia worked as a workaround so far 👌

**Comment by zaneschepke** [comment]
If someone could try this debug build and let me know if it fixes the problem that would be great! https://cloud.zaneschepke.com/s/de2g3zJEAYd59nr

**Comment by freijon** [comment]
The link doesn't work for me. Connection timed out.

**Comment by zaneschepke** [comment]
> The link doesn't work for me. Connection timed out.

Gotcha! Try this one: https://github.com/zaneschepke/wgtunnel/releases/tag/speed-testing

**Comment by freijon** [comment]
I still have the issue with this build.

**Comment by zaneschepke** [comment]
> I still have the issue with this build.

Thanks for testing. I'm sorry I would test myself but I am unable to reproduce this issue. Is there any improvement with this build? https://github.com/zaneschepke/wgtunnel/releases/edit/speed-testinv2

**Comment by freijon** [comment]
Thank you! I gave that new build a try, unfortunately the issue persists. However, while testing this I figured out that I only have the issue on mobile data, not on Wifi (as mentioned before by @hudcap). I even went back and tested both test builds again. It happens only with mobile data, Wifi works fine. Hope that helps.

**Comment by zaneschepke** [comment]
> Thank you! I gave that new build a try, unfortunately the issue persists. However, while testing this I figured out that I only have the issue on mobile data, not on Wifi (as mentioned before by @hudcap). I even went back and tested both test builds again. It happens only with mobile data, Wifi works fine. Hope that helps.

Thank you so much! This should help me narrow down what is wrong with the build. Thank you again.

**Comment by Gertjuh** [comment]
Same issue here on a Poco F3. I can confirm that the problem is only on mobile data. I am also using the original Wireguard App and that is working fine.

**Comment by nillebor** [comment]
I have the same problem with a POCO F3 with LineageOS 21. The speediest shows me download values of 0.3–0.6 Mbit. If I activate the kernel modules, I get the 40 Mbit in the download again. For me, the problem seems to be only with the download. The upload is the same with and without kernel module.

Another small problem: when the tunnel is activated, the VPN symbol (key) is not displayed in the status bar.

**Comment by hb9eue** [comment]
Same here:

- App Version 3.5.1
- Fairphone 4 5G - Android 13
- Measuring with fast.com via Mobile Network (no Wifi)

I get around 300 kb/s

When turning on the Amnezia Backend this doubles to about 600 kb/s

When using WireGuard for Android v1.0.20231018 I get about 3Mit/s

**Comment by Plopaplopa** [comment]
Same here, Xiaomi Redmi 13C. Stuck at max 300ko/s since a few days. My phone is up to date.
My girlfriend does not have the issue (Redmi 9S).
I tried to reinstall the app and reconfigure the tunnel, reboot my Wireguard server, etc. I also tried using the official Wireguard app, it works fine.

Then I finally ended here, and tried the Amnezia solution, it fixed the problem.

Anyway, good luck with this bug, and thanks for this amazing app :)

**Comment by hb9eue** [comment]
I'm wondering about those which solved the issue by enabling amnezia. My WG Server is a Mikrotik, it does not support Amnezia, I only get about double the speed when enabling Amnezia which is still slow, especially when using apps which load map tiles like google or OSM. So sould enabling Amnezia on the client still solve the issue?

**Comment by JanTijhof** [comment]
I had the same issue as everyone above. 
Getting the apk for 3.4.8 from the Fdroid repo got the connection back to the previous speeds.

Device: Mi9T
Rom: Evolution X on android 14
Version: 3.5.1
Network: mobile data

**Comment by zaneschepke** [comment]
Unfortunately, I am still unable to reproduce this issue. If anyone is able to test this version to see if it fixes the problem that would be great!

https://github.com/zaneschepke/wgtunnel/releases/tag/speed-test3

**Comment by crimsonthunder** [comment]
Testing it right now. Thank you for your continuous work!

**Comment by Nosebear** [comment]
For me the latest testversion isn't working. Still slow. Version 3.4.8 works fine. Hope you can figure out a solution. It's a great app. Thumbs up!

**Comment by crimsonthunder** [comment]
Sadly same for me. Still getting slowdowns ☹️

**Comment by JanTijhof** [comment]
No luck for me either. Greatly appreciating the work and test releases however!

**Comment by DerDreschner** [comment]
I have the same issue. The wireguard server is running on an Ubiquiti DreamMachine Pro and the app/wireguard tunnel on my Fairphone 4 is running in userspace (visible under support -> mode). Maybe that's important...?

**Comment by zaneschepke** [comment]
Does this issue only happen in userspace mode or does it it happen with kernel mode as well?

**Comment by zaneschepke** [comment]
I was unable to figure out what part of my fork on the original wireguard android lib was causing this slowness issue. 

Considering Amenzia is already working properly, I change the default backend in 3.5.2 to be Amnezia and removed the userspace toggle option from the settings. Amnezia already has backwards compatibility and the same speeds as the original, official lib. I think this should effectively mark this issue as resolved. If there is any other strange behavior around this slowness issue, just let me know and we can reopen. Thank you all for your patience with this one!

**Comment by crimsonthunder** [comment]
Thank you for all your work!

**Comment by DerDreschner** [comment]
> Does this issue only happen in userspace mode or does it it happen with kernel mode as well?

Uhm, I don't know how to use Wireguard on my phone with kernel mode. Or was that question asked towards the people who have no issues?

But as I didn't had any issues with Amnezia, I think it's a good option to use it as default backend. I looked into your code as well, but I have no experience with developing for Android to investigate it with a debug build on my phone.

**Comment by zaneschepke** [comment]
> > Does this issue only happen in userspace mode or does it it happen with kernel mode as well?
> 
> Uhm, I don't know how to use Wireguard on my phone with kernel mode. Or was that question asked towards the people who have no issues?
> 
> But as I didn't had any issues with Amnezia, I think it's a good option to use it as default backend. I looked into your code as well, but I have no experience with developing for Android to investigate it with a debug build on my phone.

Np! You have to have a rooted phone to use kernel mode. I was just asking in general if anyone was experiencing this issue in kernel mode.

**Comment by dvalter** [comment]
@zaneschepke I believe to have an idea about this issue.

wireguard-go [d0bc03](https://github.com/WireGuard/wireguard-go/commit/d0bc03c707974a84a672716c718f99fab49e7eb8) commit has enabled GSO/GRO, which may cause significant packet loss on mobile network for some reason.

On the only phone (Pixel 5a) I've tested, this very bug wad present in wireguard-android as well, and reverting GRO/GSO in wireguard-go had fixed it for me.

**Comment by zaneschepke** [comment]
Thanks for this! Ill give this a try.

**Comment by mr-scrawley** [comment]
I think I'm still affected by this in newer versions. I even tried re-adding my connections in 3.5.2 to no avail. I have not tried this in newer versions, yet, but still experience "slowness" on 3.6.0. That should not be necessary anyway to get the new default backend, anyway, right?

Am I alone in this or can anyone else confirm they are still affected before more effort is put into what could be an individual issue on my end?

**Comment by mr-scrawley** [comment]
> I think I'm still affected by this in newer versions (reverting to 3.4.8 fixes it). I even tried re-adding my connections in 3.5.2 to no avail. I have not tried this in newer versions, yet, but still experience "slowness" on 3.6.0. That should not be necessary anyway to get the new default backend, anyway, right?
> 
> Am I alone in this or can anyone else confirm they are still affected before more effort is put into what could be an individual issue on my end?

Apparently the issue now has went away. After unsuccesfully trying every new release for some time, then sticking with the old version for some time I am happy to report that the current version does not have the issue.

**But: I had to disable the "IPv4 hostname resolution" toggle**. With that enabled, the slowness persists. With it disabled, speeds are fine :)
